### PR TITLE
[MLIR][GPUToLLVMSPV] Use `llvm.func` attributes to convert `gpu.shuffle`

### PR DESCRIPTION
**Pull request by victor-eds**
Use `llvm.func`'s `intel_reqd_sub_group_size` attribute instead of SPIR-V environment attributes in the `gpu.shuffle` conversion pattern. This metadata is needed to check the semantics of the operation are supported, i.e., it has a constant width and its value is equal to the sub-group size.

As the pass also converts `gpu.func` to `llvm.func`, adding a discardable attribute of name `intel_reqd_sub_group_size` attribute to the latter is enough for this pattern to work.

We no longer have a notion of "default" sub-group size, so this attribute needs to be set in the parent function for `gpu.shuffle` operations to be converted.

Drop dependency on the SPIR-V dialect as we no longer require creating attributes from this dialect to lower `gpu.shuffle` instances.

### CODE REVIEW

**Comment by victor-eds** [comment]
This addresses our concerns discussed in https://github.com/llvm/llvm-project/pull/90972#discussion_r1589431887

File: mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir [hunk 1]
```diff
@@ -383,7 +308,20 @@ gpu.module @shuffles attributes {
...
+  }
+}
+
+// -----
+
+// Check attaching a discardable attribute to a `gpu.func` hosting a `gpu.shuffle` operation works,
+// i.e., the attribute is propagated.
+
+gpu.module @shuffles_ {
+  gpu.func @propagated_size(%val: i32, %id: i32) attributes {intel_reqd_sub_group_size = 16 : i32} {
```
Comment (FMarno) [inline mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir:321]:
```suggestion
  gpu.func @propagated_size(%val: i32, %id: i32) attributes {llvm.intel_reqd_sub_group_size = 16 : i32} {
```
I thought discardable attributes needed a prefix

**Review by FMarno — COMMENTED** [review]
just a wee thing

File: mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir (hunk 1 above)
Comment (victor-eds) [reply mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir:321]: This isn't a discardable attribute from the `llvm` dialect, but an attribute that will be propagated in the conversion to `llvm.func` and understood as the `llvm.func` op attribute.

This is a use-case I am not interested in, but an example of how you could have a parent function, convert that to LLVM and then convert the `gpu.shuffle` operation. Do you think it'd be better to drop this case?

Comment (FMarno) [reply mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir:321]: if it's not your use case then I maybe wouldn't worry about it. Add what you need, right?
I will say that you have called it a discardable attribute in the comment above.
I could be wrong about this but I was under the impression that there are inherent and discardable attributes, and this seems to be neither.

**Review by FMarno — APPROVED** [review]

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**